File: src/stable/drivers/kobuki_driver/sensors/pose3d.cpp
#include "pose3d.h"

Pose3D::Pose3D(KobukiManager *kobuki)
{
    this->kobuki = kobuki;
}

int Pose3D::setPose3DData(const jderobot::Pose3DDataPtr&  pose3dData,
                             const Ice::Current&)
{

}


jderobot::Pose3DDataPtr Pose3D::getPose3DData(const Ice::Current&)
{

    mutex.lock();
    jderobot::Pose3DDataPtr pose3dData(new jderobot::Pose3DData());
    pose3dData->x = kobuki->getRobotX();
    pose3dData->y = kobuki->getRobotY();
    pose3dData->q0 = 0.0;
    pose3dData->q1 = 0.0;
    pose3dData->q2 = 0.0;
    pose3dData->q3 = 0.0;
    mutex.unlock();
    return pose3dData;
}

File: src/drivers/MAVLinkServer/Pose3D.py
__author__ = 'AeroCano'

import jderobot, time, threading

lock = threading.Lock()

class Pose3DI(jderobot.Pose3D):

    def __init__(self,_x,_y,_z,_h,_q0,_q1,_q2,_q3):

        self.x = _x
        self.y = _y
        self.z = _z
        self.h = _h
        self.q0 = _q0
        self.q1 = _q1
        self.q2 = _q2
        self.q3 = _q3

    def setPose3DData(self,_x,_y,_z,_h,_q0,_q1,_q2,_q3):

        lock.acquire()
        #print 'adquired S'

        self.x = _x
        self.y = _y
        self.z = _z
        self.h = _h
        self.q0 = _q0
        self.q1 = _q1
        self.q2 = _q2
        self.q3 = _q3

        #print 'released S'
        lock.release()

        return 0



    def getPose3DData(self, current=None):

        time.sleep(0.05) #50 ms rate to tx Pose3D

        lock.acquire()
        #print 'adquired G'

        data = jderobot.Pose3DData()
        data.x = self.x
        data.y = self.y
        data.z = self.z
        data.h = self.h
        data.q0 = self.q0
        data.q1 = self.q1
        data.q2 = self.q2
        data.q3 = self.q3

        #print 'released G'
        lock.release()

        return data